core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -134,151 +102,137 @@ public Object visitPrepareCommand(TxInvocationContext ctx, PrepareCommand comman
    }
 
-   private boolean needRemoteLocks(InvocationContext ctx, Object key, FlagAffectedCommand command)
-         throws Throwable {
-      boolean needBackupLocks = ctx.isOriginLocal() && (!isLockOwner(key) || isStateTransferInProgress());
-      boolean needRemoteLock = false;
-      if (needBackupLocks && !command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL)) {
-         final TxInvocationContext txContext = (TxInvocationContext) ctx;
-         LocalTransaction localTransaction = (LocalTransaction) txContext.getCacheTransaction();
-         needRemoteLock = !localTransaction.getAffectedKeys().contains(key);
-         if (!needRemoteLock) {
-            if (trace)
-               log.tracef("We already have lock for key %s, skip remote lock acquisition", key);
-         }
-      } else {
-         if (trace)
-            log.tracef("Don't need backup locks %s", needBackupLocks);
-      }
-      return needRemoteLock;
+   private Object handleRemoteLockCommand(TxInvocationContext context, LockControlCommand command)
+         throws InterruptedException {
+      lockAllOrRegisterBackupLock(context, command.getKeys(), getLockTimeoutMillis(command));
+      return invokeNext(context, command);
    }
 
-   private boolean isLockOwner(Collection<?> keys) {
-      for (Object key : keys) {
-         if (!isLockOwner(key)) {
-            return false;
-         }
-      }
-      return true;
+   private boolean skipLockingForKey(FlagAffectedCommand command, TxInvocationContext<?> ctx, Object key) {
+      return !ctx.isOriginLocal() ||
+             hasSkipLocking(command) ||
+             isKeysAlreadyLocked(ctx, key);
+   }
+
+   private boolean skipLockingForKeys(FlagAffectedCommand command, TxInvocationContext<?> ctx,
+         Collection<?> keys) {
+      return !ctx.isOriginLocal() ||
+             hasSkipLocking(command) ||
+             areAllKeysAlreadyLocked(ctx, keys);
    }
 
-   private boolean isStateTransferInProgress() {
-      return stateTransferManager != null && stateTransferManager.isStateTransferInProgress();
+   private boolean isLocalOnly(FlagAffectedCommand command) {
+      return command.hasAnyFlag(FlagBitSets.CACHE_MODE_LOCAL);
    }
+
+   private Object invokeLockCommandAndContinue(Collection<?> keys, FlagAffectedCommand command,
+         TxInvocationContext<?> ctx, int topologyId) throws InterruptedException {
+      //we need to acquire locks
+      LockControlCommand lcc = cf.buildLockControlCommand(keys, command.getFlagsBitSet(), ctx.getGlobalTransaction());
+      lcc.setTopologyId(topologyId);
+      return makeStage(handleLocalLockCommand(ctx, lcc))
+            .thenApply(ctx, command, (rCtx, rCommand, rv) -> invokeNext(rCtx, rCommand));
+   }
+
+   private Object handleLocalLockCommand(TxInvocationContext context, LockControlCommand command)
+         throws InterruptedException {
+      //TODO blocking method. change it to non-blocking? [minor, only blocks local commands]
```

Comment:
Maybe you've missed my previous comment here: I believe that this part must not be blocking as it would block `xxxAsync` methods.

Reply:
that is another issue. I won't fix it in this PR

Reply:
☹️ Please create a JIRA and mention it in the comment then.